Pier boundary condition stiffness

barischrooneyj opened the pull request:
@rozsasarpi I have moved your PR here. I made a few changes, the TCL file isn't quite as the example that you had laid out, but it is mostly setup to get there. The last few bits shouldn't be so bad.

Perhaps take a look through the changes (apart from the `.tcl` file that is there).

If you uncomment line 214 in `bridge_sim/sim/run/opensees/build/d3/__init__.py` then duplicate nodes for each pier will be placed into the `.tcl` file. That function is also where the remaining changes need to be made, currently I iterate through each pier and through the bottom nodes (and duplicates) of that pier.

If you `checkout` this branch and then run the following you can see the changes made in the `.tcl` file. This is also useful to do as you make any changes to the code.

```
rm -rf generated-data && poetry run python pr-example.py
cp generated-data/bridge-705-10,0-healthy/opensees/bridge-705-10,0-healthy-params=5,-0,-100.tcl og-temp.tcl
git diff
```

barischrooneyj commented:
A few notes:

- `BuildContext.get_uniaxial_material` may be useful.
- I noticed how bad some of the variable naming is. Some of this is from the very beginning of the project. I think it's best to avoid making too many changes at once and keep the PR simple.
- I made `LineSpringSupport` an argument of `Support` (which is really a `Pier` and will be renamed as such). One reason is that every pier needs to have boundary conditions, so it's useful to have it attached there. And secondly, it forces a user of the library to think "this pier needs boundary conditions", in contrast if all of the `LineSpringSupport`s are given as arguments to `Bridge` then the user is not forced to think of this association and this check would have to be done explicitly somewhere else.
- The main thing changed from your PR was the calculation of the duplicate nodes. The `BuildContext.get_node` function was modified to take an argument `nth_node` which returns the nth node at that position. (Still needs to be updated for a potential bug, but is good enough for right now, I will update it before the PR is closed).

rozsasarpi commented:
@barischrooneyj Thanks for clear description. I made some steps (the changes are pushed to the branch):

* most changes are in `opensees_pier_boundary_conditions`; see the `TODO`s there
* uniaxial material OS commands are generated
* zerolength elements OS commands are generated though the element numbering is incorrect. I think we need an element id counter method in `BuildContext` i.e. `new_e_id`
* `uniaxial_materials` has duplicate elements, keep only the unique ones

Could you check the changes in the code and look into the remaining `TODO`s?

barischrooneyj commented:
> Could you check the changes in the code and look into the remaining TODOs?
Yep, I'll get back to you by tomorrow.

barischrooneyj commented:
Hi @rozsasarpi 

- `uniaxialMaterial` and `element zeroLength` are now inserted into the TCL file without duplicate IDs, I have added two  TODOs in `opensees_pier_boundary_conditions`. One is just a potential method of making the code a little shorter, and the other is a request for a comment.
- Currently only `stiffness_z_rotation` is used, what about the other 5 stiffness values?
- `get_node` has seen a bug fix, previously it would have failed if you call it with `nth_node=2` if it had not yet been called with `nth_node=1`.
- Regarding the "close stiffness values": I guess the problem is that for two or more stiffness values (floating point numbers) are very close so that the difference could be ignored in practice but multiple `uniaxialMaterial` materials get created? So this doesn't really affect the running of the code currently but it would be desirable to round to slightly above machine precision or something like that? Perhaps an issue should be opened for this, this rounding of values is something I encountered also with calculating positions of nodes and I think it should be revisited.
- I changed some arguments from `n_i` to `ni_id` etc. to be consistent with how I have been referring to node IDs in other parts of the code and also since we have `s_id` `e_id` etc. It also saves a second in thinking about whether I have a node vs a node ID.

barischrooneyj commented:
> There are nodes that are very close, why is this?

I'm not sure but have opened an issue and will look into it. #189

rozsasarpi commented on bridge_sim/sim/run/opensees/build/d3/__init__.py:
```diff
@@ -179,13 +181,143 @@ def opensees_fixed_abutment_nodes(
+            all_bottom_nodes.append(bottom_nodes)
+            all_dupe_bottom_nodes.append(dupe_bottom_nodes)
+
+    # TODO: If the information provided by having these nested lists will not be
```
@barischrooneyj I'm not sure if a set would work. As far as I know a set does not have an order of its elements but the order information is relevant (at least in the current code), see for example the loop starting at [line 263](https://github.com/barischrooneyj/bridge-sim/pull/188/files#diff-6b4283943bb654184358a714e509de15R263). However, the outer list of the nested lists does not seem to be needed.

rozsasarpi commented on bridge_sim/sim/run/opensees/build/d3/__init__.py:
```diff
@@ -179,13 +181,143 @@ def opensees_fixed_abutment_nodes(
+    # needed. Then the code can be shortened by creating sets on line 207
+    # instead of lists, and adding directly to these.
+    all_dupe_bottom_nodes_flat = flatten(all_dupe_bottom_nodes, Node)
+    all_uniaxial_materials_flat = set(flatten(all_uniaxial_materials, UniaxialMaterial))
```
The only disadvantage of converting the list to a set is that the order information is lost so the `tcl` file is harder to read. However, this is a tiny issue we can live with.

rozsasarpi commented on bridge_sim/sim/run/opensees/build/d3/__init__.py:
```diff
@@ -179,13 +181,143 @@ def opensees_fixed_abutment_nodes(
+            # diff = np.diff(bottom_node_coords, axis=0)
+            # diff_dist = np.sum(diff ** 2, axis=1)
+            #
+            # TODO: Comment on diff_dist and node_dist.
```
Here is my graphical description:

```
n_1     n_2     n_3  n_4
o ====== o ==== o == o

sought for distances (node_dist):
|------|-----------|-------|---|
```

`diff_dist` is a poorly named variable that contains the difference of the `z` coordinates of subsequent nodes. This assumes that the nodes are in order, probably we should not assume this but we should enforce it by sorting them (minding if the sorting would mess up things).

`node_dist` for each node is contains the length of the interval that is defined as follows: interval start point: at halfway to the subsequent point on the left; interval end point: at halfway to the subsequent point on the right. Actually my code is not correct, there should be a `/2` for each `diff_dist`.

The `numpy` based code is making the above described calculation in an efficient vectorized fashion.

rozsasarpi commented on bridge_sim/sim/run/opensees/build/d3/__init__.py:
```diff
@@ -179,13 +181,143 @@ def opensees_fixed_abutment_nodes(
+            # TODO: Comment on diff_dist and node_dist.
+            diff_dist = np.diff(bottom_node_coords[:,2])
+            node_dist = np.sum(
+                np.vstack((np.hstack((0, diff_dist)), np.hstack((diff_dist, 0)))),
```
```suggestion
                np.vstack((np.hstack((0, diff_dist/2)), np.hstack((diff_dist/2, 0)))),
```

rozsasarpi commented:
@barischrooneyj 

* I think (haven't checked it but I'm >95% sure) that the current assignment of material ids (`$matTag` in OS) is incorrect, the different materials are numbered independently, i.e. `nDMaterial ElasticOrthotropic` and `uniaxialMaterial Elastic`. They are both materials and I think OS would expect their numbering to be consistent, e.g. there cannot be a  `nDMaterial ElasticOrthotropic` and `uniaxialMaterial Elastic` both with `matTag = 1` I think every time you create a new material the `new_e_id` method should be called to get the right number. I do not understand yet how the `nDMaterial ElasticOrthotropic` receives its id integer.

* > Currently only stiffness_z_rotation is used, what about the other 5 stiffness values?

It would be great to have the other 5 stiffness values as well, I was just going for a quick implementation/solution. For bridge 705 we have a strong feeling that the rotation about the `z` axis is the (only) relevant degree of freedom to be constrained with a spring. If you have the time and it would be great if you could add the other directions as well. Do not add a duplicate node if there is not at least one spring stiffness different from zero. I expect that OS would not like that, i.e. having a node not connecting to the rest of the structure.

* I just realized that one component is missing from the pier boundary condition generation. All duplicate nodes at the bottom of the pier (only those that are created the second) should have all of their degrees of freedom fixed (fix 9999 1 1 1 1 1 1), this way they are anchored to the "earth" and can really exert some resistance against rotation (or other displacements). Could you add this?

* "close stiffness values", yes it was a reminder due to that reason that you gave (floating point accuracy). Based on your comment I agree that it is not a problem and it is fine to have multiple materials even if from a practical point of view it would not be necessary. It think this has a very low priority, the current way of dealing with it is good.

* Renaming variables: sure, good changes.

barischrooneyj commented:
> I think (haven't checked it but I'm >95% sure) that the current assignment of material ids ($matTag in OS) is incorrect 

I noticed this after OpenSees shouted at me, it's fixed already.

> It would be great to have the other 5 stiffness values as well

In my view as package maintainer I think this is necessary. The previous less powerful API `fix_z_rotation` etc. was functional, and this PR should not introduce replacement features that are not fully implemented. https://blog.codinghorror.com/the-broken-window-theory/

> All duplicate nodes at the bottom of the pier (only those that are created the second, so one member of each twin) should have all of their degrees of freedom fixed ... Could you add this?

Yep, I can add this, I have some time on Friday

Roadmap to getting this merged:
- `stiffness_z_rotation` e.g. "I was just going for a quick implementation/solution."
- finish 5 remaining stiffness properties
- handle pier settlement (previously the nodes where automatically freed for Y translation for a pier undergoing settlement)
- test suite passing

rozsasarpi commented:
@barischrooneyj I run the code and checked the `.tcl` file. It looks good to me, it has everything what we need for the specific use case. In the coming days I'll propose and prepare a simple case to very the computations.

barischrooneyj commented:
Will be mostly offline for next two weeks. However would be great to merge this this month. I reckon the first point on the "roadmap" above is the vast majority of the work, and can fly through the rest.

barischrooneyj commented:
Remembered one thing that is undocumented that you might run into. Currently the stiffness parameters are not taken into account when computing a path of where to store results from simulation.

As a temporary workaround you can change the bridge.name attribute to save/load results from a different path. I think this is the default right now (that simulations results for one `Bridge` are loaded/saved based on `Bridge.name` + loading parameters). The reason being that this is easier to maintain (instead of taking every single `Bridge` parameter into account when calculating a file path), and just pushed the burden onto the user.

barischrooneyj commented:
> @barischrooneyj I run the code and checked the `.tcl` file. It looks good to me, it has everything what we need for the specific use case. In the coming days I'll propose and prepare a simple case to very the computations.

@rozsasarpi any luck on verifying the `stiffness_z_rotation`? I can do the remaining parameters next.

barischrooneyj commented:
As an aside: my motivation for this project is to have a nice portfolio project, something that is technical, substantial and polished. I will try to finish  https://github.com/barischrooneyj/bridge-sim/milestone/14 this year and then start tackling the paper feedback the next few months as milestone v0.0.2. Received some great feedback from the JOSS review, https://github.com/openjournals/joss-reviews/issues/2496#issuecomment-691732204 and the following comments are relevant.
